daemon/SecretsImpl/secrets.cpp

```diff
@@ -499,69 +504,149 @@ bool Daemon::ApiImpl::SecretsRequestQueue::generateKeyData(
     kdfParams.setSalt(salt);
     kdfParams.setOutputKeySize(256);
 
-    Sailfish::Crypto::CryptoPlugin *cplugin = m_autotestMode
-            ? m_controller->crypto()->plugins().value(Sailfish::Crypto::CryptoManager::DefaultCryptoPluginName + QLatin1String(".test"))
-            : m_controller->crypto()->plugins().value(Sailfish::Crypto::CryptoManager::DefaultCryptoPluginName);
-
-    if (cplugin == Q_NULLPTR) {
-        qCWarning(lcSailfishSecretsDaemon) << "Unable to find default crypto plugin for key initialization";
+    // attempt to find the crypto plugin to use to perform key derivation.
+    const QMap<QString, Sailfish::Crypto::CryptoPlugin*> cplugins = m_controller->crypto()->plugins();
+    Sailfish::Crypto::CryptoPlugin *cplugin = cipherPluginName.isEmpty() ? Q_NULLPTR : cplugins.value(cipherPluginName);
+    if (cplugin == Q_NULLPTR && !cipherPluginName.isEmpty()) {
+        qCWarning(lcSailfishSecretsDaemon) << "Unable to find parameter cipher plugin to generate keys:" << cipherPluginName;
         return false;
     }
 
-    Sailfish::Crypto::Result bkdbKeyResult = cplugin->generateKey(
-                keyTemplate,
-                Sailfish::Crypto::KeyPairGenerationParameters(),
-                kdfParams,
-                QVariantMap(),
-                &bookkeepingdbKey);
+    // attempt to use the plugin specified in the environment variable
+    const QString envCPluginName = QString::fromLocal8Bit(qgetenv("SAILFISH_SECRETSD_MASTERLOCK_CRYPTOPLUGIN"));
```

Comment:
`SAILFISH_SECRETSD_MASTERLOCK_CRYPTOPLUGIN` should not be hardcoded here. I suggest to `#define` it in some header file. Also some documentation wouldn't hurt.

Reply:
good point, done